packages/alfa-map/src/map.ts

```diff
@@ -48,59 +48,63 @@ export class Map<K, V> implements Collection.Keyed<K, V> {
     return this._size === 0;
   }
 
-  public forEach(callback: Callback<V, void, [K]>): void {
+  public forEach(callback: Callback<V, void, [key: K]>): void {
     Iterable.forEach(this, ([key, value]) => callback(value, key));
   }
 
-  public map<U>(mapper: Mapper<V, U, [K]>): Map<K, U> {
+  public map<U>(mapper: Mapper<V, U, [key: K]>): Map<K, U> {
     return new Map(this._root.map(mapper), this._size);
   }
 
-  public flatMap<L, U>(mapper: Mapper<V, Map<L, U>, [K]>): Map<L, U> {
+  public apply<U>(mapper: Map<K, Mapper<V, U>>): Map<K, U> {
+    return mapper.flatMap((mapper) => this.map(mapper));
+  }
+
+  public flatMap<L, U>(mapper: Mapper<V, Map<L, U>, [key: K]>): Map<L, U> {
     return this.reduce(
       (map, value, key) => map.concat(mapper(value, key)),
       Map.empty<L, U>()
     );
   }
 
-  public reduce<R>(reducer: Reducer<V, R, [K]>, accumulator: R): R {
+  public flatten<K, V>(this: Map<K, Map<K, V>>): Map<K, V> {
```

Comment:
`flatMap` and `flatten` would also benefit from some tests. Notably to see what happens with key collisions.

Reply:
Agreed! I've copied the remarks from `Map#distinct()` to both methods:

> As the order of maps is undefined, it is also undefined which keys are deleted when duplicate values are encountered.